### PR TITLE
Rewrite in TypeScript

### PR DESCRIPTION
**Pull request by staltz**
I sometimes use JS, sometimes use TS, and I don't intend to push for TS everywhere in the SSB stack, but secret-stack is definitely one that could direly use TS's capabilities. Here are some strong reasons to use TypeScript:

### Pros

- We need `secret-stack` to be rigorously checked for security
  - This part of the SSB stack is critical for security, and we want to be rigorous about `null` checks, `undefined` checks, safety against typos, etc. Some linters give you some safety, but TypeScript is ultimately a very strong linter. I know at least of one serious bug in the past in this library that would have hardly slipped by if TypeScript were used. That's the main motivation to write in TS: to make this library defensive against bugs (which can be also security vulnerabilities).
- Types can aid in defining valid states, documenting expected values, etc
- Other languages have types, so defining them here help people implement SSB in other languages

### Cons (and counterarguments)

- People might have difficulty reading and contributing to TS code, compared to JS code
  - Counterargument: we haven't had many people reading nor contributing to secret-stack, so this wouldn't change much
  - Counterargument: this library doesn't seem to need to evolve much, besides bug fixes, so it's not critical that it must be the most readable code
  - Counterargument: Dominic's code style was already arguable quite difficult to read and contribute to, and while I rewrite this I can gradually make it simpler to understand
- Compilation step before testing, or before publishing
  - Counterargument: I set up package.json scripts so that it always compiles before either testing or publishing
  - Counterargument: it takes just 1 second on my computer to compile

### How to review this PR

- Read this commit first: [move source code to directory src/](https://github.com/ssbc/secret-stack/commit/b22779d8c1ff2187723ab985921993244452fe33)
  - This simply moves the **JS** files to a new place
  - Done because (in the next commit) the compiled JS will live in `lib/` and the TS will live in `src/`
  - `lib/` ends up in the npm package but not in git
- Then read this commit: [rewrite src/ files in basic TypeScript](https://github.com/ssbc/secret-stack/commit/f44f192fe2de88647dbe1391c5a9e9a9d33f03e1)
  - My approach here was simply rename `.js` to `.ts` and toss some `any` (this is the most loose type in TypeScript) whenever the TS compiler complained
  - I will submit another PR as soon as this one is merged, where I use "proper" TypeScript, so far this PR only sets up the TS project but doesn't use any fancy TS features

### CODE REVIEW

File: src/core.ts [hunk 1]
```diff
@@ -1,36 +1,36 @@
...
   return o && typeof o === 'object' && !Array.isArray(o)
 }
 
 var isArray = Array.isArray
 
-function toBase64 (s) {
+function toBase64 (s: Buffer | string) {
   if (isString(s)) return s
-  else s.toString('base64') // assume a buffer
+  else return s.toString('base64') // assume a buffer
```
Comment (staltz) [inline src/core.ts:25]: Here's an example of bug that TS caught. This function was returning undefined, when it should return a string always

**Comment by christianbundy** [comment]
Thanks for writing this patch!

I'd run `npm install` to fix the lockfile and push the results, but otherwise I wouldn't stand in the way of this change.

**Comment by staltz** [comment]
Oh yeah I'll update that. I was using pnpm locally

**Comment by staltz** [comment]
Ok, updated PR to update package-lock.json too. I wouldn't want to merge this myself, as I think reviewing secret-stack is very important in case a sneaky mistake slipped in.

**Comment by mixmix** [comment]
I've not reviewed this and don't know if I have capacity atm, but I'm very excited about this @staltz . I'd love to get a few people together soon and talk about what we're each focusing on.
BTW I'd also consider this a maintenance task, so eligible for scuttlebutt-maintenance grant funding if that would be of service <3

**Comment by staltz** [comment]
> BTW I'd also consider this a maintenance task, so eligible for scuttlebutt-maintenance grant funding if that would be of service <3

Good point! Mind if I invoice the open collective with ~20€/h hourly rate? (that's the same I use for Manyverse, and `~` means approximate because the OC uses a different currency than euro)

**Comment by arj03** [comment]
I've been talking with @staltz about having a call as well. Just to catch up and see where we are at.

**Comment by mixmix** [comment]
Ok who's going to read through this closely?
Also up for funding.

I noticed I was reading through this carefully on my phone but not
intending to do it all.
Fwiw it looks fairly easy to review (and I don't even know ts well)

**Comment by mixmix** [comment]
I'm gonna try asking specific people.
Hey @arj03 can you review this please (at a rate of $50 NZD an hour)

_If you can't take it on, can you ask another specific person plz_

File: src/core.ts [hunk 2]
```diff
@@ -83,21 +83,21 @@ module.exports = {
...
 
-    var timeoutInactivity
+    var timeoutInactivity: number
     if (opts.timers && !isNaN(opts.timers.inactivity)) {
       timeoutInactivity = opts.timers.inactivity
     }
     // if opts.timers are set, pick a longer default
     // but if not, set a short default (as needed in the tests)
-    timeoutInactivity = timeoutInactivity || (opts.timers ? 600e3 : 5e3)
+    timeoutInactivity = timeoutInactivity! || (opts.timers ? 600e3 : 5e3)
```
Comment (arj03) [inline src/core.ts:96]: Why is this needed?

File: src/index.ts [hunk 3]
```diff
@@ -1,8 +1,7 @@
-'use strict'
-var Api = require('./api')
+import Api = require('./api')
 
 // config must have appKey
-module.exports = function (config) {
+export = function SecretStack (config: any) {
```
Comment (arj03) [inline src/index.ts:4]: ❤️

**Comment by arj03** [comment]
Did a quick review. There was one minor thing I stumpled on, otherwise looks good. 

Havn't really done any typescript coding but as I've been reading ssb-conn code as well, the most important part for me is that the code is more readable especially to new people, so I'm fully behind converting this to ts.

File: src/core.ts (hunk 2 above)
Comment (staltz) [reply src/core.ts:96]: Oh, `!` at the end of an identifier are a TypeScript-only feature. It may look weird since it reminds the `!` "not" operator.

`timeoutInactivity` was declared a `number` but it may not be assigned (if the `if` doesn't trigger), so TypeScript was warning that I can't "use" the value of `timeoutInactivity` yet. `timeoutInactivity!` tells TypeScript to ignore the risks of it not being assigned yet, and just "trust me". It turns out to compile to just `timeoutInactivity`.

There's an alternative solution to this which is declaring `timeoutInactivity` as `number | undefined`. But I didn't want that because code below the lines 96 will always see a `number`, never `undefined`, so I didn't want the rest of the code to handle undefined if it's never undefined. So essentially lines 90–96 comprise the declaration **and assignment** of `timeoutInactivity` and we are sure that after line 96 it's always a `number`.

Comment (arj03) [reply src/core.ts:96]: Alright, thanks for the explanation. Makes sense.

**Comment by arj03** [comment]
Merge away. Quite interested to see the next patch :-)

**Comment by mixmix** [comment]
thanks @arj03 
Can you please put an expense in to https://opencollective.com/scuttlebutt-maintenance in the next couple of days. Really appreciate your doing this important maintenance work

(staltz has already put an expense in for some of this)

**Comment by arj03** [comment]
Thanks. It's a good idea to also pay people to review, but as I'm not in any kind of money short @staltz can you expense an extra hour on the next bill and consider that an donation to manyverse.

**Comment by mixmix** [comment]
Love it.

Sometime I'd like to talk about how we each relate to money. I think it
could be really relevant to maintenance
